### DISCUSSION TITLE
pdnsutil/API: Add unique IDs to everything

### DISCUSSION THREAD

**Discussion by TCB13** [PowerDNS/pdns#13318] [Ideas]
Hello,

This isn't a bug report, this is just something that crosses my mind every single time I use the API or pdnsutil:

Why can't we have unique IDs for records exposed in the API? And modify / delete records using them? Also the API would be way more RESTful.

The current approach works however:

1) It makes it hard sometimes to deal with duplications and whatnot;
2) Sometimes forces people into building middleware systems that keep a separate database that hold a copy of all records just so you can have proper IDs and expose things as true REST... > this isn't good from a performance / complexity standpoint and becomes problematic when the PDNS API simply doesn't deal well with large number of sequential requests.

Exposing the unique ID of records on the API would make everyone's life easier. Why can't the API simply print out the `ID` column of the `records` table and/or use it?

Thank you.

**Comment by Habbie** [comment]
We made a deliberate choice to identify everything by name where possible. We felt this would actually make for a cleaner API. While we have no plans to change this, I am curious about your points (1) and (2) - and perhaps we can think about smaller changes that would make your life easier. Can you tell us more about the problems you run into with the current API? What's the problem with duplication?

**Reply by TCB13** [reply]
Okay,

Consider you're building a web frontend to PDNS, such as the great https://github.com/PowerDNS-Admin/PowerDNS-Admin, now you're pulling all the records from the PDNS API and showing them on your UI. 

We've some types of record that can have multiple values for the same name, eg. TXT:
````
example.org TXT "validationkey=1234"
example.org TXT "validationkey=56789"
````

Now if a user wants to delete the second record `validationkey=56789` I'm forced to jump through hoops to get it done. The usual approach is:

1. Query PDNS to get the zone: ` "GET", "zones/" . $name`
2. Parse the entire thing
3. Search into `rrsets` > `records` for some record which `content` equals `"validationkey=56789"`´ and filter out that record;
4. Build a new `"PATCH", "zones/" . $name` query with a `"changetype" => "REPLACE",` and send back the remaining records.

**Suggestion:** every record has an ID. A simple `DELETE /zones/zone-name/record-id` query would be way more efficient than having both the client and PDNS parse entire `rrsets` for differences...

I might be wrong, but as I see it this scales really poorly. Right now if a user wants to delete a single record you're essentially requiring the WebUI or some middleware to keep track of an entire `rrset` and send it again to the PDNS API without the record to delete.

Even if you have a look at PowerDNS-Admin you'll see that it takes a 1-3 seconds to delete a record. If we had IDs it could've, most likely, been done in milliseconds.

Further on, the current `Deleting a RRset` query isn't even that useful:
````
PATCH /api/v1/servers/localhost/zones/example.org.
{"rrsets": [{"name": "test.example.org.", "type": "A", "changetype": "DELETE"}]}
````

Because the same thing can be achieved in a more predictable way by using an empty `REPLACE`:
````
PATCH /api/v1/servers/localhost/zones/example.org.
{"rrsets": [{"name": "test.example.org.", "type": "A", "ttl": 3600, "changetype": "REPLACE", "records": [] }
````

Thank you.

**Reply by Habbie** [reply]
What if we added a way to GET a single RRset?

**Reply by TCB13** [reply]
So you suggestion is something like:

```` 
GET /api/v1/servers/localhost/zones/example.org./test.example.org
````

That would return an rsset with the `name` = `test.example.org.`  Is that it?

That would be useful, however the main "issue" still remains, in order to remove a record from an rsset I've to fetch it, compute what needs to be removed and then issue a `REPLACE` query. This is why an ID for each record would make things easier.

It would also be nice to see `DELETE` with the same logic:

```` 
DELETE /api/v1/servers/localhost/zones/example.org./test.example.org
````

So we could delete entire rssets without having to specify a `type`.

**Reply by Habbie** [reply]
if you don't specify a type, it's not an RRset - it's all RRsets for a name

**Reply by TCB13** [reply]
Okay that's fair but what about the records inside an rsset? Anwyays, what's the logic behind the concept of an RRset? Why not simply have records?

**Reply by klaus-nicat** [reply]
For example all RRs of an RRset must have the same TTL. So there are properties which are common for the whole RRset.

**Reply by TCB13** [reply]
Why not simply just have records? I don't see the benefit, just extra overhead and extra steps to get things into the API. Even when we look at the MySQL database there's no concept of RRsets, just records...

**Comment by klaus-nicat** [comment]
Do you want the ID only for deleting? Or also for updating a single RR of a RRset? And do you also want to add single RRs to an existing RRset?

**Reply by TCB13** [reply]
It would be good to have a way / ID to delete and update both the RRsets and the records inside them.
